Fix StorageCluster status image nil pointer dereference issue

malayparida2000 opened the pull request:
Move initializeImagesStatus() call to after all potential Client.Update() operations to prevent API server responses from overwriting initialized status fields.

This fixes nil pointer panics in ceph/noobaa ensure functions that occur when finalizer/annotation updates overwrite in-memory status during the first reconcile loop.
Ref-https://issues.redhat.com/browse/DFBUGS-2902

malayparida2000 commented:
Without fix Panic is seen during the 1st reconcile

`"panic":"runtime error: invalid memory address or nil pointer dereference","panicGoValue":"\"invalid memory address or nil pointer dereference\"","stacktrace":"goroutine 1400 [running]:\nk8s.io/apimachinery/pkg/util/runtime.logPanic({0x270eda0, 0xc000bf6060}, {0x1fed040, 0x3981bb0})\n\t/src/remote_source/app/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:132 +0xbc\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile.func1()\n\t/src/remote_source/app/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:108 +0x112\npanic({0x1fed040?, 0x3981bb0?})\n\t/usr/lib/golang/src/runtime/panic.go:792 +0x132\ngithub.com/red-hat-storage/ocs-operator/v4/controllers/storagecluster.(*ocsNoobaaSystem).ensureCreated(0xc0001ead00?, 0xc0001ead00, 0xc001bb8588)`

With this fix the panic is no longer. visible after the creation of storageCluster

malayparida2000 commented on controllers/storagecluster/reconcile.go:
```diff
@@ -387,6 +384,10 @@ func (r *StorageClusterReconciler) reconcilePhases(
 
+	// Initialize the StatusImages section after all potential Client.Update() calls
+	// to prevent them from being overwritten by API server responses
+	r.initializeImagesStatus(instance)
```
>did you already check after this and before we actually update the status of storagecluster there is no op overwriting the in-memory obj?

Yes so StorageCluster CR is our own object so we never update it's spec. We only update either the metadata(in this case finalizer & annotation), Or the status. And I checked that all the update calls on the StorageCluster object are done before this line. Hence I have added a comment to touch the status section only after all potential update calls.
>maybe it's better to transfer the concern where we actually set these values so that we'll be shielded against any other future introduction of code b/n this & status update line?

There can be 2 ways of that I can think of
1. Add a not nill check before assigning the value, But that would not be correct as even though the panic would not happen but the status.images section would stay empty till another reconcile happens.
2. After initializing the status image immediately call the Client.Status.Update(), But that would add a extra call to the API server in each reconcile and that would revert https://github.com/red-hat-storage/ocs-operator/pull/885/commits/2ccffe1e27dbdaa795a9eae1c7f2e3fbde679d79, Which might have unknown side effects.

leelavg commented:
/hold

There is an ongoing thread, @malayparida2000 the change and description is way off now.

malayparida2000 commented:
> /hold
> 
> There is an ongoing thread, @malayparida2000 the change and description is way off now.

That was some crazy mess up from both me & @iamniting , Thanks a lot @leelavg for catching it.

Actually, we were on a call and discussed the problem in detail. And we arrived at the conclusion that a more proper fix would be to return & requeue the reconcile anytime a client.Update() is called. 
I implemented the change & pushed the commit, but forgot to change the PR desc.
Nitin immediately approved the changes, but later we noticed the issue of the mismatch of the PR desc and the ongoing thread. 
But we saw the unit tests are also failing, as we were already running late, we concluded the meet, hoping the PR would not be merged due to the failing unit test. Hence the mess up.

Well, now I have fixed the PR desc as well as the failing tests as well. Please take a look at the approach now.
